File: js/lib/arena.js
// SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-Clockwork-Commercial
// Copyright (c) 2026 Sam Aaron
/**
 * arena.js — the arena's table of contents, as JavaScript reads it.
 *
 * Mirrors src/clockwork_arena.h: the header at the front of the arena, its
 * entries, the region ids and what each entry's geometry words mean. Read
 * BY ID, never by position — a reader that walked the table positionally was
 * the drift this replaces (BufferLayout, read as uint32View[n]).
 *
 * `readArena(buffer, base)` returns the header, a `region(id)` lookup, and
 * `constants` — the flat object the rest of the JS has always consumed
 * (IN_BUFFER_START, METRICS_SIZE, SHM_SCOPE_MAX_SCOPES, …), now derived from
 * the table rather than declared beside it.
 */

export const ARENA_MAGIC = 0x43574152;      // 'CWAR'
export const ARENA_VERSION = 1;
export const ARENA_PUBLISHED = 1;
const ENTRY_WORDS = 16;                     // sizeof(ClockworkArenaEntry) / 4
const HEADER_WORDS = 16;                    // the fixed fields before the entries

export const OWNER = Object.freeze({ CLOCKWORK: 1, GUEST: 2, CLIENT: 3, HOST: 4 });

export const REGION = Object.freeze({
  CONTROL: 1, METRICS: 2, NATIVE_STATS: 3, CLOCK_STATE: 4, CLOCK_ANCHORS: 5,
  SAMPLE_CLOCK: 6, CHANNEL_MAP: 7, NODE_ID_COUNTER: 8, IN_RING: 9, OUT_RING: 10,
  NRT_OUT_RING: 11, AUDIO_TAPS: 12, CLIENT_SLOTS: 13, GUEST_CONFIG: 14,
  GUEST_WINDOW: 15, SCOPE: 16, GUEST_PERSIST: 17, TRACK_TAPS: 18,
});

// Geometry word indices, per region (ClockworkArenaGeom).
export const GEOM = Object.freeze({
  RING_MAX_MESSAGE: 0, RING_MESSAGE_MAGIC: 1, RING_PADDING_MAGIC: 2, RING_PADDING_MARKER: 3, RING_HEADER_BYTES: 4,
  METRICS_FIELDS: 0,
  ANCHOR_NTP_START: 0, ANCHOR_DRIFT: 1, ANCHOR_GLOBAL: 2,
  TAPS_SLOTS: 0, TAPS_SLOT_BYTES: 1, TAPS_HEADER_BYTES: 2, TAPS_FRAMES: 3, TAPS_CHANNELS: 4, TAPS_SAMPLE_RATE: 5,
  SCOPE_SLOTS: 0, SCOPE_HEADER_BYTES: 1, SCOPE_SLOT_BYTES: 2, SCOPE_SLOT_HEADER: 3, SCOPE_RING_FRAMES: 4, SCOPE_CHANNELS: 5,
  TRACK_SLOTS: 0, TRACK_SLOT_BYTES: 1, TRACK_SLOT_HEADER: 2, TRACK_RING_FRAMES: 3, TRACK_CHANNELS: 4, TRACK_FIRST_INDEX: 5,
  SLOTS_COUNT: 0, SLOTS_SLOT_BYTES: 1, SLOTS_HEADER_BYTES: 2, SLOTS_STRUCTS_OFF: 3, SLOTS_STRUCTS_BYTES: 4,
  SLOTS_STACK_OFF: 5, SLOTS_STACK_BYTES: 6,
});

/**
 * Read the header at `base` in `buffer`. Throws if it is not a published
 * arena of a version this reader knows, so a caller never proceeds on a
 * table it half-understands.
 *
 * @param {ArrayBuffer|SharedArrayBuffer} buffer
 * @param {number} base byte offset of the arena
 */
export function readArena(buffer, base) {
  if (base % 4 !== 0) throw new Error(`arena base ${base} is not 4-byte aligned`);
  const words = new Uint32Array(buffer, base, HEADER_WORDS);
  const [magic, version, headerBytes, instanceId, arenaBytes, blockBytes, guestOffset, guestBytes,
         entryCount, entryBytes, state] = words;
  if (magic !== ARENA_MAGIC) throw new Error(`not a clockwork arena (magic ${magic.toString(16)})`);
  if (version !== ARENA_VERSION) throw new Error(`arena version ${version}, this reader knows ${ARENA_VERSION}`);
  if (state !== ARENA_PUBLISHED) throw new Error('arena not yet published');
  if (entryBytes !== ENTRY_WORDS * 4) throw new Error(`arena entry shape ${entryBytes} bytes, expected ${ENTRY_WORDS * 4}`);

  const table = new Uint32Array(buffer, base + HEADER_WORDS * 4, entryCount * ENTRY_WORDS);
  const entries = new Map();
  for (let i = 0; i < entryCount; i++) {
    const at = i * ENTRY_WORDS;
    const id = table[at];
    if (id === 0) throw new Error(`arena entry ${i} is empty`);
    const offset = table[at + 1], bytes = table[at + 2];
    if (offset < headerBytes || offset + bytes > arenaBytes) throw new Error(`arena region ${id} lies outside the arena`);
    entries.set(id, { id, offset, bytes, owner: table[at + 3], geom: Array.from(table.subarray(at + 4, at + ENTRY_WORDS)) });
  }
  const region = (id) => {
    const e = entries.get(id);
    if (!e) throw new Error(`arena has no region ${id}`);
    return e;
  };

  return {
    header: { magic, version, headerBytes, instanceId, arenaBytes, blockBytes, guestOffset, guestBytes, entryCount },
    entries,
    region,
    has: (id) => entries.has(id),
    constants: constantsFrom(region, entries, { arenaBytes, blockBytes, guestOffset, guestBytes, version, instanceId }),
  };
}

/**
 * The flat constants object the JS runtime consumes, derived from the table.
 * The names are the ones shared_memory.h has always used, so a reader of
 * either side recognises them; the values come from THIS engine's table.
 */
function constantsFrom(region, entries, h) {
  // Optional: an arena with no track taps has no entry, which is zero slots.
  const track = entries.get(REGION.TRACK_TAPS) || { offset: 0, bytes: 0, geom: [] };
  const inRing = region(REGION.IN_RING), outRing = region(REGION.OUT_RING), nrt = region(REGION.NRT_OUT_RING);
  const control = region(REGION.CONTROL), metrics = region(REGION.METRICS), window = region(REGION.GUEST_WINDOW);
  const anchors = region(REGION.CLOCK_ANCHORS), clock = region(REGION.CLOCK_STATE), taps = region(REGION.AUDIO_TAPS);
  const nodeId = region(REGION.NODE_ID_COUNTER), config = region(REGION.GUEST_CONFIG), scope = region(REGION.SCOPE);
  const sample = region(REGION.SAMPLE_CLOCK), persist = region(REGION.GUEST_PERSIST), map = region(REGION.CHANNEL_MAP);
  const slots = region(REGION.CLIENT_SLOTS);
  return {
    ARENA_VERSION: h.version,
    ARENA_INSTANCE_ID: h.instanceId,
    CLOCKWORK_BLOCK_SIZE: h.blockBytes,
    GUEST_REGION_START: h.guestOffset,
    GUEST_REGION_SIZE: h.guestBytes,
    TOTAL_BUFFER_SIZE: h.arenaBytes,

    IN_BUFFER_START: inRing.offset, IN_BUFFER_SIZE: inRing.bytes,
    OUT_BUFFER_START: outRing.offset, OUT_BUFFER_SIZE: outRing.bytes,
    NRT_OUT_BUFFER_START: nrt.offset, NRT_OUT_BUFFER_SIZE: nrt.bytes,
    MAX_MESSAGE_SIZE: inRing.geom[GEOM.RING_MAX_MESSAGE],
    MESSAGE_MAGIC: inRing.geom[GEOM.RING_MESSAGE_MAGIC],
    PADDING_MAGIC: inRing.geom[GEOM.RING_PADDING_MAGIC],
    RING_PADDING_MARKER: inRing.geom[GEOM.RING_PADDING_MARKER],
    MESSAGE_HEADER_SIZE: inRing.geom[GEOM.RING_HEADER_BYTES],

    CONTROL_START: control.offset, CONTROL_SIZE: control.bytes,
    METRICS_START: metrics.offset, METRICS_SIZE: metrics.geom[GEOM.METRICS_FIELDS] * 4,
    SHM_WINDOW_START: window.offset, SHM_WINDOW_SIZE: window.bytes,
    NTP_START_TIME_START: anchors.offset + anchors.geom[GEOM.ANCHOR_NTP_START], NTP_START_TIME_SIZE: 8,
    DRIFT_OFFSET_START: anchors.offset + anchors.geom[GEOM.ANCHOR_DRIFT], DRIFT_OFFSET_SIZE: 4,
    GLOBAL_OFFSET_START: anchors.offset + anchors.geom[GEOM.ANCHOR_GLOBAL], GLOBAL_OFFSET_SIZE: 4,
    CLOCK_STATE_START: clock.offset, CLOCK_STATE_SIZE: clock.bytes,
    SHM_AUDIO_START: taps.offset, SHM_AUDIO_TOTAL_SIZE: taps.bytes,
    SHM_AUDIO_SLOTS: taps.geom[GEOM.TAPS_SLOTS],
    SHM_AUDIO_OUT_SLOT: 0,   // CLOCKWORK_TAP_OUT: what left for the device
    SHM_AUDIO_IN_SLOT: 1,    // CLOCKWORK_TAP_IN: what arrived from it
    SHM_AUDIO_SLOT_SIZE: taps.geom[GEOM.TAPS_SLOT_BYTES],
    SHM_AUDIO_HEADER_SIZE: taps.geom[GEOM.TAPS_HEADER_BYTES],
    SHM_AUDIO_FRAMES: taps.geom[GEOM.TAPS_FRAMES],
    SHM_AUDIO_CHANNELS: taps.geom[GEOM.TAPS_CHANNELS],
    SHM_AUDIO_SAMPLE_RATE: taps.geom[GEOM.TAPS_SAMPLE_RATE],
    NODE_ID_COUNTER_START: nodeId.offset, NODE_ID_COUNTER_SIZE: nodeId.bytes,
    GUEST_CONFIG_START: config.offset, GUEST_CONFIG_SIZE: config.bytes,
    SHM_SCOPE_START: scope.offset, SHM_SCOPE_TOTAL_SIZE: scope.bytes,
    SHM_SCOPE_MAX_SCOPES: scope.geom[GEOM.SCOPE_SLOTS],
    SHM_SCOPE_HEADER_SIZE: scope.geom[GEOM.SCOPE_HEADER_BYTES],
    SHM_SCOPE_SLOT_SIZE: scope.geom[GEOM.SCOPE_SLOT_BYTES],
    SHM_SCOPE_SLOT_HEADER_SIZE: scope.geom[GEOM.SCOPE_SLOT_HEADER],
    SHM_SCOPE_RING_FRAMES: scope.geom[GEOM.SCOPE_RING_FRAMES],
    SHM_SCOPE_CHANNELS: scope.geom[GEOM.SCOPE_CHANNELS],
    // The engine's track taps: scope slots numbered after the guest's.
    SHM_TRACK_TAPS_START: track.offset,
    SHM_TRACK_TAPS_SIZE: track.bytes,
    SHM_TRACK_TAPS_SLOTS: track.geom[GEOM.TRACK_SLOTS] || 0,
    SHM_TRACK_TAPS_FIRST_INDEX: track.geom[GEOM.TRACK_FIRST_INDEX] || scope.geom[GEOM.SCOPE_SLOTS],
    // Every slot a client can number: the guest's, then the track taps.
    SHM_SCOPE_SLOT_COUNT: scope.geom[GEOM.SCOPE_SLOTS] + (track.geom[GEOM.TRACK_SLOTS] || 0),
    SAMPLE_CLOCK_START: sample.offset, SAMPLE_CLOCK_SIZE: sample.bytes,
    GUEST_PERSIST_START: persist.offset, GUEST_PERSIST_SIZE: persist.bytes,
    CHANNEL_MAP_START: map.offset, CHANNEL_MAP_SIZE: map.bytes,
    CLIENT_SLOTS_START: slots.offset, CLIENT_SLOTS_SIZE: slots.bytes,
    CLIENT_SLOT_COUNT: slots.geom[GEOM.SLOTS_COUNT],
    CLIENT_SLOT_SIZE: slots.geom[GEOM.SLOTS_SLOT_BYTES],
    CLIENT_SLOTS_HEADER_SIZE: slots.geom[GEOM.SLOTS_HEADER_BYTES],
    CLIENT_SLOT_STRUCTS_OFFSET: slots.geom[GEOM.SLOTS_STRUCTS_OFF],
    CLIENT_SLOT_STRUCTS_SIZE: slots.geom[GEOM.SLOTS_STRUCTS_BYTES],
    CLIENT_SLOT_STACK_OFFSET: slots.geom[GEOM.SLOTS_STACK_OFF],
    CLIENT_SLOT_STACK_SIZE: slots.geom[GEOM.SLOTS_STACK_BYTES],
  };
}

File: rust/clockwork-abi/src/arena.rs
// SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-Clockwork-Commercial
// Copyright (c) 2026 Sam Aaron
//! The arena's table of contents (`src/clockwork_arena.h`, `docs/ARENA.md`).
//!
//! The first bytes of every arena: a magic, a version, and a table naming
//! each region — where, how big, who writes it, and the geometry a reader
//! walks it by. Looked up BY ID, never by position. A client with only the
//! base pointer finds everything from here; so does a host on the lanes ABI,
//! through [`crate::lanes::clockwork_arena_header`].

pub const CLOCKWORK_ARENA_MAGIC: u32 = 0x4357_4152; // 'CWAR'
pub const CLOCKWORK_ARENA_VERSION: u32 = 1;
pub const CLOCKWORK_ARENA_HEADER_BYTES: u32 = 4096;
pub const CLOCKWORK_ARENA_MAX_ENTRIES: u32 = 48;
pub const CLOCKWORK_ARENA_GEOM_WORDS: usize = 12;
pub const CLOCKWORK_ARENA_LAYING_OUT: u32 = 0;
pub const CLOCKWORK_ARENA_PUBLISHED: u32 = 1;

/// Who READS a region: a bit set in the last geometry word of every entry
/// (`geom::AUDIENCE`). Owner says who writes; this says who it is for.
/// 0 is nobody but the owner — or a writer from before the word existed.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
#[repr(transparent)]
pub struct ClockworkArenaAudience(pub u32);
impl ClockworkArenaAudience {
    /// A data contract for any reader in a client process, read by hand.
    pub const PUBLISHED: Self = Self(1);
    /// The guest, through a pointer clockwork hands it.
    pub const GUEST: Self = Self(2);
    /// The host.
    pub const HOST: Self = Self(4);
    /// The command plane: the client ABI only, never by hand.
    pub const TRANSPORT: Self = Self(8);
}

/// Who writes a region.
#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub struct ClockworkArenaOwner(pub u32);
impl ClockworkArenaOwner {
    pub const CLOCKWORK: Self = Self(1);
    pub const GUEST: Self = Self(2);
    pub const CLIENT: Self = Self(3);
    pub const HOST: Self = Self(4);
}

/// The regions. Stable numbers.
#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub struct ClockworkArenaRegion(pub u32);
impl ClockworkArenaRegion {
    pub const CONTROL: Self = Self(1);
    pub const METRICS: Self = Self(2);
    pub const NATIVE_STATS: Self = Self(3);
    pub const CLOCK_STATE: Self = Self(4);
    pub const CLOCK_ANCHORS: Self = Self(5);
    pub const SAMPLE_CLOCK: Self = Self(6);
    pub const CHANNEL_MAP: Self = Self(7);
    pub const NODE_ID_COUNTER: Self = Self(8);
    pub const IN_RING: Self = Self(9);
    pub const OUT_RING: Self = Self(10);
    pub const NRT_OUT_RING: Self = Self(11);
    pub const AUDIO_TAPS: Self = Self(12);
    pub const CLIENT_SLOTS: Self = Self(13);
    pub const GUEST_CONFIG: Self = Self(14);
    pub const GUEST_WINDOW: Self = Self(15);
    pub const SCOPE: Self = Self(16);
    pub const GUEST_PERSIST: Self = Self(17);
    /// Scope-stream slots the engine writes from plugin-track returns,
    /// numbered after the guest's scope slots in one index space.
    pub const TRACK_TAPS: Self = Self(18);
}

/// The two audio taps, as slot indices into `AUDIO_TAPS`: what left for the
/// device this block, and what arrived from it. Written by clockwork inside
/// every tick, on every host.
pub const CLOCKWORK_TAP_OUT: u32 = 0;
pub const CLOCKWORK_TAP_IN: u32 = 1;

/// Geometry word indices, per region (`ClockworkArenaGeom`).
pub mod geom {
    pub const RING_MAX_MESSAGE: usize = 0;
    pub const RING_MESSAGE_MAGIC: usize = 1;
    pub const RING_PADDING_MAGIC: usize = 2;
    pub const RING_PADDING_MARKER: usize = 3;
    pub const RING_HEADER_BYTES: usize = 4;
    pub const METRICS_FIELDS: usize = 0;
    pub const ANCHOR_NTP_START: usize = 0;
    pub const ANCHOR_DRIFT: usize = 1;
    pub const ANCHOR_GLOBAL: usize = 2;
    pub const TAPS_SLOTS: usize = 0;
    pub const TAPS_SLOT_BYTES: usize = 1;
    pub const TAPS_HEADER_BYTES: usize = 2;
    pub const TAPS_FRAMES: usize = 3;
    pub const TAPS_CHANNELS: usize = 4;
    pub const TAPS_SAMPLE_RATE: usize = 5;
    pub const SCOPE_SLOTS: usize = 0;
    pub const SCOPE_HEADER_BYTES: usize = 1;
    pub const SCOPE_SLOT_BYTES: usize = 2;
    pub const SCOPE_SLOT_HEADER: usize = 3;
    pub const SCOPE_RING_FRAMES: usize = 4;
    pub const SCOPE_CHANNELS: usize = 5;
    pub const TRACK_SLOTS: usize = 0;
    pub const TRACK_SLOT_BYTES: usize = 1;
    pub const TRACK_SLOT_HEADER: usize = 2;
    pub const TRACK_RING_FRAMES: usize = 3;
    pub const TRACK_CHANNELS: usize = 4;
    pub const TRACK_FIRST_INDEX: usize = 5;
    pub const SLOTS_COUNT: usize = 0;
    pub const SLOTS_SLOT_BYTES: usize = 1;
    pub const SLOTS_HEADER_BYTES: usize = 2;
    pub const SLOTS_STRUCTS_OFF: usize = 3;
    pub const SLOTS_STRUCTS_BYTES: usize = 4;
    pub const SLOTS_STACK_OFF: usize = 5;
    pub const SLOTS_STACK_BYTES: usize = 6;
    /// EVERY region: the audience (`ClockworkArenaAudience` bits), in the
    /// last word. A region's own geometry stops short of it.
    pub const AUDIENCE: usize = super::CLOCKWORK_ARENA_GEOM_WORDS - 1;
}

// The audience word is the last geometry word, and no region's own geometry
// reaches it. Pinned at compile time, where the C header pins the same.
const _: () = assert!(geom::AUDIENCE == CLOCKWORK_ARENA_GEOM_WORDS - 1);
const _: () = assert!(geom::SLOTS_STACK_BYTES < geom::AUDIENCE);
const _: () = assert!(geom::TRACK_FIRST_INDEX < geom::AUDIENCE);

/// One region. 64 bytes.
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct ClockworkArenaEntry {
    pub id: u32,
    pub offset: u32,
    pub bytes: u32,
    pub owner: u32,
    pub geom: [u32; CLOCKWORK_ARENA_GEOM_WORDS],
}

impl ClockworkArenaEntry {
    /// Who this region is for (`ClockworkArenaAudience` bits); 0 from a
    /// writer that did not say.
    pub fn audience(&self) -> u32 {
        self.geom[geom::AUDIENCE]
    }
    /// A data contract a client may read by hand.
    pub fn published(&self) -> bool {
        self.audience() & ClockworkArenaAudience::PUBLISHED.0 != 0
    }
}

/// The header, at arena offset 0.
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct ClockworkArenaHeader {
    pub magic: u32,
    pub version: u32,
    pub header_bytes: u32,
    pub instance_id: u32,
    pub arena_bytes: u32,
    pub block_bytes: u32,
    pub guest_offset: u32,
    pub guest_bytes: u32,
    pub entry_count: u32,
    pub entry_bytes: u32,
    pub state: u32,
    /// The block's published run is `[header_bytes, block_published_end)`;
    /// the rest of the block is transport. 0: not stated by the writer.
    pub block_published_end: u32,
    /// The guest's published run is `[guest_offset, guest_published_end)`;
    /// the rest is the guest's own. 0: not stated.
    pub guest_published_end: u32,
    pub reserved: [u32; 3],
    pub entries: [ClockworkArenaEntry; CLOCKWORK_ARENA_MAX_ENTRIES as usize],
}

impl ClockworkArenaHeader {
    /// The entry for a region, if the arena has one.
    pub fn find(&self, id: ClockworkArenaRegion) -> Option<&ClockworkArenaEntry> {
        let n = (self.entry_count as usize).min(CLOCKWORK_ARENA_MAX_ENTRIES as usize);
        self.entries[..n].iter().find(|e| e.id == id.0)
    }

    /// True when the arena has region `id` and it is a data contract a
    /// client may read by hand. False for a region the arena lacks, one
    /// that is transport or private, and one from a writer that never said.
    pub fn published(&self, id: ClockworkArenaRegion) -> bool {
        self.find(id).is_some_and(|e| e.published())
    }

    /// True when this is a published table of a version this crate knows,
    /// fitting in `bytes`, with every region inside the arena it describes.
    /// `Err` names the first failure, in the C header's words.
    pub fn check(&self, bytes: u32) -> Result<(), &'static str> {
        if bytes < core::mem::size_of::<Self>() as u32 { return Err("no room for an arena header"); }
        if self.magic != CLOCKWORK_ARENA_MAGIC { return Err("not a clockwork arena"); }
        if self.version != CLOCKWORK_ARENA_VERSION { return Err("arena version this reader does not know"); }
        if self.state != CLOCKWORK_ARENA_PUBLISHED { return Err("arena not yet published"); }
        if self.header_bytes < core::mem::size_of::<Self>() as u32 { return Err("header reservation smaller than the header"); }
        if self.entry_bytes != core::mem::size_of::<ClockworkArenaEntry>() as u32 { return Err("entry shape this reader does not know"); }
        if self.entry_count > CLOCKWORK_ARENA_MAX_ENTRIES { return Err("more entries than the table holds"); }
        if self.arena_bytes > bytes { return Err("arena larger than the memory it is in"); }
        if self.block_bytes > self.arena_bytes { return Err("clockwork block larger than the arena"); }
        if self.guest_offset > self.arena_bytes || self.guest_bytes > self.arena_bytes - self.guest_offset {
            return Err("guest region outside the arena");
        }
        for e in &self.entries[..self.entry_count as usize] {
            if e.id == 0 { return Err("an empty entry inside the count"); }
            if e.offset < self.header_bytes { return Err("a region overlaps the header"); }
            if e.offset > self.arena_bytes || e.bytes > self.arena_bytes - e.offset {
                return Err("a region runs past the arena");
            }
        }
        // The published runs, when the writer stated them: each inside its
        // half, every published region inside its run, everything else after.
        if self.block_published_end != 0 || self.guest_published_end != 0 {
            if self.block_published_end < self.header_bytes || self.block_published_end > self.block_bytes {
                return Err("the block's published run runs past the block");
            }
            if self.guest_published_end < self.guest_offset
                || self.guest_published_end > self.guest_offset + self.guest_bytes
            {
                return Err("the guest's published run is outside the guest region");
            }
            for e in &self.entries[..self.entry_count as usize] {
                let end = if e.offset >= self.guest_offset { self.guest_published_end } else { self.block_published_end };
                if e.published() && e.offset + e.bytes > end {
                    return Err("a published region outside the published run");
                }
                if !e.published() && e.offset < end {
                    return Err("an unpublished region inside the published run");
                }
            }
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn entry(id: u32, offset: u32, bytes: u32, audience: u32) -> ClockworkArenaEntry {
        let mut geom = [0u32; CLOCKWORK_ARENA_GEOM_WORDS];
        geom[geom::AUDIENCE] = audience;
        ClockworkArenaEntry { id, offset, bytes, owner: ClockworkArenaOwner::CLOCKWORK.0, geom }
    }

    // A small arena: header, a published run of two regions, a transport run
    // of one; a guest half with one published and one private region.
    fn header() -> ClockworkArenaHeader {
        let mut h = ClockworkArenaHeader {
            magic: CLOCKWORK_ARENA_MAGIC, version: CLOCKWORK_ARENA_VERSION,
            header_bytes: CLOCKWORK_ARENA_HEADER_BYTES, instance_id: 0,
            arena_bytes: 12288, block_bytes: 8192, guest_offset: 8192, guest_bytes: 4096,
            entry_count: 5, entry_bytes: core::mem::size_of::<ClockworkArenaEntry>() as u32,
            state: CLOCKWORK_ARENA_PUBLISHED,
            block_published_end: 6144, guest_published_end: 10240, reserved: [0; 3],
            entries: [entry(0, 0, 0, 0); CLOCKWORK_ARENA_MAX_ENTRIES as usize],
        };
        h.entries[0] = entry(ClockworkArenaRegion::METRICS.0, 4096, 1024, ClockworkArenaAudience::PUBLISHED.0);
        h.entries[1] = entry(ClockworkArenaRegion::CLOCK_STATE.0, 5120, 1024,
                             ClockworkArenaAudience::PUBLISHED.0 | ClockworkArenaAudience::GUEST.0);
        h.entries[2] = entry(ClockworkArenaRegion::IN_RING.0, 6144, 2048, ClockworkArenaAudience::TRANSPORT.0);
        h.entries[3] = entry(ClockworkArenaRegion::GUEST_WINDOW.0, 8192, 2048, ClockworkArenaAudience::PUBLISHED.0);
        h.entries[4] = entry(ClockworkArenaRegion::GUEST_PERSIST.0, 10240, 2048, ClockworkArenaAudience::GUEST.0);
        h
    }

    #[test]
    fn a_well_laid_out_table_passes_and_says_who_is_published() {
        let h = header();
        assert_eq!(h.check(12288), Ok(()));
        assert!(h.published(ClockworkArenaRegion::METRICS));
        assert!(h.published(ClockworkArenaRegion::GUEST_WINDOW));
        assert!(!h.published(ClockworkArenaRegion::IN_RING));
        assert!(!h.published(ClockworkArenaRegion::GUEST_PERSIST));
        assert!(!h.published(ClockworkArenaRegion::SCOPE)); // absent
        assert_eq!(h.find(ClockworkArenaRegion::CLOCK_STATE).unwrap().audience() & ClockworkArenaAudience::GUEST.0, 2);
    }

    #[test]
    fn a_published_region_past_its_run_is_refused() {
        let mut h = header();
        h.entries[2].geom[geom::AUDIENCE] = ClockworkArenaAudience::PUBLISHED.0;
        assert_eq!(h.check(12288), Err("a published region outside the published run"));
    }

    #[test]
    fn an_unpublished_region_inside_the_run_is_refused() {
        let mut h = header();
        h.entries[0].geom[geom::AUDIENCE] = ClockworkArenaAudience::TRANSPORT.0;
        assert_eq!(h.check(12288), Err("an unpublished region inside the published run"));
    }

    #[test]
    fn a_boundary_outside_its_half_is_refused() {
        let mut h = header();
        h.block_published_end = 8192 + 16;
        assert_eq!(h.check(12288), Err("the block's published run runs past the block"));
        let mut h = header();
        h.guest_published_end = 8192 - 16;
        assert_eq!(h.check(12288), Err("the guest's published run is outside the guest region"));
    }

    #[test]
    fn a_writer_that_said_nothing_is_not_judged() {
        let mut h = header();
        h.block_published_end = 0;
        h.guest_published_end = 0;
        for e in h.entries.iter_mut() { e.geom[geom::AUDIENCE] = 0; }
        assert_eq!(h.check(12288), Ok(()));
        assert!(!h.published(ClockworkArenaRegion::METRICS));
    }

}
